src/rust/engine/nailgun/src/server.rs

```diff
@@ -0,0 +1,355 @@
+      // NB: This closure captures the stdio handles, and will drop/close them when it completes.
+      (nail.runner)(RawFdExecution {
+        cmd,
+        stdin_fd: stdin_handle.as_raw_fd(),
+        stdout_fd: stdout_handle.as_raw_fd(),
+        stderr_fd: stderr_handle.as_raw_fd(),
+      })
+    });
+
+    // Fully consume the stdout/stderr streams before waiting on the exit stream.
+    let stdout_stream = stdout_stream.map_ok(ChildOutput::Stdout);
+    let stderr_stream = stderr_stream.map_ok(ChildOutput::Stderr);
+    let exit_stream = exit_code_future
+      .into_stream()
+      .map(|exit_code| Ok(ChildOutput::Exit(exit_code)));
+    let output_stream = stream::select(stdout_stream, stderr_stream)
+      .chain(exit_stream)
+      .map(|res| match res {
+        Ok(o) => o,
+        Err(e) => {
+          error!("IO error interacting with the runner: {:?}", e);
+          ChildOutput::Exit(ExitCode(-1))
+        }
+      })
+      .boxed();
+
+    Ok(Child {
+      output_stream,
+      accepts_stdin,
+    })
+  }
+}
+
+impl RawFdNail {
+  ///
+  /// Returns a tuple of a readable file handle and an optional sink for nails to send stdin to.
+  ///
+  /// In the case of a TTY, the file handle will point directly to the TTY, and no stdin data will
+  /// flow over the protocol. Otherwise, it will be backed by a new anonymous pipe, and data should
+  /// be copied to the returned Sink.
+  ///
+  fn input(
+    tty_path: Option<PathBuf>,
+  ) -> Result<(Box<dyn AsRawFd + Send>, Option<impl sink::Sink<Bytes>>), io::Error> {
+    if let Some(tty_path) = tty_path {
+      Ok((Box::new(std::fs::File::open(tty_path)?), None))
+    } else {
+      let (stdin_reader, stdin_writer) = os_pipe::pipe()?;
+      let write_handle =
+        File::from_std(unsafe { std::fs::File::from_raw_fd(stdin_writer.into_raw_fd()) });
+      Ok((Box::new(stdin_reader), Some(sink_for(write_handle))))
+    }
+  }
+
+  ///
+  /// Returns a tuple of a possibly empty Stream for nails to read data from, and a writable file handle.
+  ///
+  /// See `Self::input` and the struct's rustdoc for more info on the TTY case.
+  ///
+  #[allow(clippy::type_complexity)]
+  fn output(
+    tty_path: Option<PathBuf>,
+  ) -> Result<
+    (
+      stream::BoxStream<'static, Result<Bytes, io::Error>>,
+      Box<dyn AsRawFd + Send>,
+    ),
+    io::Error,
+  > {
+    if let Some(tty_path) = tty_path {
+      let tty = std::fs::OpenOptions::new()
+        .write(true)
+        .create(false)
+        .open(tty_path)?;
+      Ok((stream::empty().boxed(), Box::new(tty)))
+    } else {
+      let (stdin_reader, stdin_writer) = os_pipe::pipe()?;
+      let read_handle =
+        File::from_std(unsafe { std::fs::File::from_raw_fd(stdin_reader.into_raw_fd()) });
+      Ok((stream_for(read_handle).boxed(), Box::new(stdin_writer)))
+    }
+  }
+
+  ///
+  /// Corresponds to `ttynames_to_env` in `nailgun_protocol.py`. See this struct's rustdocs.
```

Comment:
Does "corresponds" mean that it reads the env vars which are set in that method?

Reply:
No idea, I'm just moving this code around. I'd prefer to keep any proposed changes to `server.rs` out of scope for this PR, since this is just moving code that used to exist in `lib.rs` into a separate file.